### PR TITLE
[STM32F1] Simple (but working) Flash EEPROM

### PR DESCRIPTION
**Pull request by trouch**
Gentlemen, here is another approach to Flash EEPROM with STM32F1.
This implementation simply stores each byte into an half word.
That way we get rid of the test_index and datasize mismatch. Easy as 1+1=10 😅 

Existing implementation was buggy at several places, so I've cut the tumor at the root !

Ok... we lose half of the virtual EEPROM, but seems far enough so far.
I'm more considered by extensive write operation always on the same page.

[STM white paper](https://www.st.com/resource/en/application_note/cd00165693.pdf) suggests to use cascaded address+data byte/word, which would lead to even more extensive flash usage. 

As a later improvement, we could add a flag to switch between pages at each eeprom save. 

Here is the result on the FYSETC Cheetah (pins update attached). I'll follow up with non EEPROM related PRs later for that guy.

```
M500 
echo:Settings Stored (602 bytes; crc 43459)
ok
M504 
echo:EEPROM OK
ok
```

### CODE REVIEW

**Comment by trouch** [comment]
@tpruvot - what do you think ?

File: Marlin/src/HAL/HAL_STM32F1/persistent_store_flash.cpp [hunk 1]
```diff
@@ -67,42 +66,29 @@ bool PersistentStore::write_data(int &pos, const uint8_t *value, size_t size, ui
...
+    }
+    crc16(crc, &v, 1);
   }
-  pos += ((size + 1) & ~1); // i.e., size+(size&1), round up odd values
+  pos += size;
   return false;
 }
 
-size_t PersistentStore::capacity() { return E2END + 1; }
+size_t PersistentStore::capacity() { return EEPROM_PAGE_SIZE; }
```
Comment (tpruvot) [inline Marlin/src/HAL/HAL_STM32F1/persistent_store_flash.cpp]: unsure you should change that... btw your pins.h offset set 2 * EEPROM_PAGE_SIZE .. wanted ?
i auto reply for the 2x : yes seems required in maple (two pages seems "used")

OK I see why you made that, if E2END is not set...
Marlin\src\HAL\HAL_STM32F1\persistent_store_flash.cpp:105:45: error: 'E2END' was not declared in this scope
 size_t PersistentStore::capacity() { return E2END + 1; }

This file may need :
```
#ifndef E2END
  #define E2END (EEPROM_PAGE_SIZE-1)
#endif
```

File: Marlin/src/HAL/HAL_STM32F1/persistent_store_flash.cpp [hunk 2]
```diff
@@ -67,42 +66,29 @@ bool PersistentStore::write_data(int &pos, const uint8_t *value, size_t size, ui
...
-    i++;
-  }
-
-  // Now, write any remaining single byte
-  const uint16_t odd = size & 1;
-  if (odd) {
-    uint16_t temp = value[size - 1];
-    status = FLASH_ProgramHalfWord(pageBase + pos + i, temp);
-    if (status != FLASH_COMPLETE) return true;
+  for (int i=0; i<size; i++) {
```
Comment (tpruvot) [inline Marlin/src/HAL/HAL_STM32F1/persistent_store_flash.cpp]: Marlin/src/HAL/HAL_STM32F1/persistent_store_flash.cpp: In static member function 'static bool PersistentStore::write_data(int&, const uint8_t*, size_t, uint16_t*)':
Marlin/src/HAL/HAL_STM32F1/persistent_store_flash.cpp:69:18: warning: comparison between signed and unsigned integer expressions [-Wsign-compare]

File: Marlin/src/HAL/HAL_STM32F1/persistent_store_flash.cpp [hunk 3]
```diff
@@ -67,42 +66,29 @@ bool PersistentStore::write_data(int &pos, const uint8_t *value, size_t size, ui
...
   return false;
 }
 
 bool PersistentStore::read_data(int &pos, uint8_t* value, const size_t size, uint16_t *crc, const bool writing/*=true*/) {
-  for (uint16_t i = 0; i < size; i++) {
-    byte* accessPoint = (byte*)(pageBase + pos + i);
-    uint8_t c = *accessPoint;
-    if (writing) value[i] = c;
-    crc16(crc, &c, 1);
+  for (int i=0; i<size; i++) {
```
Comment (tpruvot) [inline Marlin/src/HAL/HAL_STM32F1/persistent_store_flash.cpp]: Marlin/src/HAL/HAL_STM32F1/persistent_store_flash.cpp: In static member function 'static bool PersistentStore::read_data(int&, uint8_t*, size_t, uint16_t*, bool)':
Marlin/src/HAL/HAL_STM32F1/persistent_store_flash.cpp:80:18: warning: comparison between signed and unsigned integer expressions [-Wsign-compare]

**Comment by tpruvot** [comment]
doesnt seems to work on our Longer Board
| Send: M500
| Recv: echo:Settings Stored (6 bytes; crc 0)
| Recv: Error:EEPROM datasize error.

if turning on DEBUG_EEPROM_READWRITE
Send: M500
Recv: Error:Field home_offset mismatch.

pins config:
+// EEPROM stored in SoC Flash (framework-arduinoststm32-maple/STM32F1/libraries/EEPROM/EEPROM.h)
+#define FLASH_EEPROM_EMULATION
+#define EEPROM_PAGE_SIZE     uint16_t(0x800)
+#define EEPROM_START_ADDRESS uint32_t(0x8000000 + 512 * 1024 - 2 * EEPROM_PAGE_SIZE)

File: Marlin/src/module/configuration_store.cpp [hunk 4]
```diff
@@ -447,36 +447,19 @@ void MarlinSettings::postprocess() {
 
 #if ENABLED(EEPROM_SETTINGS)
 
-  #define WORD_PADDED_EEPROM ENABLED(__STM32F1__, FLASH_EEPROM_EMULATION)
-
-  #if WORD_PADDED_EEPROM && ENABLED(DEBUG_EEPROM_READWRITE)
-    #define UPDATE_TEST_INDEX(VAR) (text_index += sizeof(VAR))
```
Comment (tpruvot) [inline Marlin/src/module/configuration_store.cpp]: I think this file should be kept, but there is a typo in current source code... should be "test_index"

File: Marlin/src/HAL/HAL_STM32F1/persistent_store_flash.cpp [hunk 5]
```diff
@@ -67,42 +66,29 @@ bool PersistentStore::write_data(int &pos, const uint8_t *value, size_t size, ui
...
+    uint8_t v = *(uint16_t *)(EEPROM_PAGE0_BASE + (pos + i) * 2);
```
Comment (tpruvot) [inline Marlin/src/HAL/HAL_STM32F1/persistent_store_flash.cpp:81]: this code line seems very strange to me... you put 2 bytes of data in only one byte...

**Comment by tpruvot** [comment]
Ok, was able to make it work on the U20... and the SKRmini
Here is the updated commit : https://github.com/tpruvot/Marlin/commit/1bc5c97b95ac3f265b291955ecebb58886613f57

**Comment by trouch** [comment]
Thanks @tpruvot for the feedbacks.
Let me answers your questions quick 
1. We store 1 byte into 2 bytes on the flash - EEPROM size is then the half of reserved Flash.
2. `0xAB` is wrote as a 2 bytes `0x00AB`
3. Reading 2 bytes casted into single byte strip out MSB.

So ... `0xAB ==> write ==> 0x00AB ==> read ==>  0xAB`

File: Marlin/src/module/configuration_store.cpp (hunk 4 above)
Comment (trouch) [reply Marlin/src/module/configuration_store.cpp]: Yeah that typo was obvious, but there was also others bugs !

1. comparing `uint16_t MarlinSettings::datasize() { return sizeof(SettingsData); }` with eeprom_index will always fail at https://github.com/MarlinFirmware/Marlin/blob/bugfix-2.0.x/Marlin/src/module/configuration_store.cpp#L2039

2. in your original code, the odd write should be at `pos + (i*2)` https://github.com/MarlinFirmware/Marlin/blob/bugfix-2.0.x/Marlin/src/HAL/HAL_STM32F1/persistent_store_flash.cpp#L85

I haven't been able to fix your approach after several hours of debug. So I came down to this new approach.

File: Marlin/src/HAL/HAL_STM32F1/persistent_store_flash.cpp (hunk 5 above)
Comment (trouch) [reply Marlin/src/HAL/HAL_STM32F1/persistent_store_flash.cpp:81]: MSB is stripped out with implicit uint16 => uint8 casting;

File: Marlin/src/module/configuration_store.cpp (hunk 4 above)
Comment (tpruvot) [reply Marlin/src/module/configuration_store.cpp]: well i didnt touch this file on my branches, and that works..
maybe just some serial debug info

Comment (trouch) [reply Marlin/src/module/configuration_store.cpp]: Ahah, I'd bet that.
1. Comparing unpadded struct size with padded index cannot work.

2. Let's do it manual 
```
...
size = 5
...
i = 4 
WriteHalfWord(K + (i*2)) = WriteHalfWord(K+8)
i= 5
// loop exit - write odd byte
WriteHalfWord(K + i) = WriteHalfWord(K+5)
...
```

File: Marlin/src/HAL/HAL_STM32F1/persistent_store_flash.cpp (hunk 5 above)
Comment (tpruvot) [reply Marlin/src/HAL/HAL_STM32F1/persistent_store_flash.cpp:81]: indeed
```
0007f0c0  ff ff ff ff ff ff ff ff  56 00 36 00 38 00 00 00  |........V.6.8...|
0007f0d0  ac 00 89 00 01 00 c8 00  00 00 00 00 00 00 c8 00  |................|
0007f0e0  00 00 00 00 00 00 64 00  00 00 00 00 00 00 b8 00  |......d.........|
0007f0f0  0b 00 00 00 00 00 20 00  4e 00 00 00 00 00 00 00  |...... .N.......|
0007f100  00 00 a0 00 42 00 00 00  00 00 a0 00 42 00 00 00  |....B.......B...|
0007f110  00 00 c8 00 43 00 00 00  00 00 c4 00 42 00 00 00  |....C.......B...|
0007f120  00 00 48 00 43 00 00 00  00 00 48 00 43 00 00 00  |..H.C.....H.C...|
0007f130  00 00 c8 00 42 00 00 00  00 00 c8 00 41 00 00 00  |....B.......A...|
0007f140  00 00 48 00 43 00 00 00  00 00 fa 00 43 00 00 00  |..H.C.......C...|
0007f150  00 00 48 00 43 00 00 00  00 00 00 00 00 00 00 00  |..H.C...........|
0007f160  00 00 00 00 00 00 00 00  00 00 20 00 41 00 00 00  |.......... .A...|
0007f170  00 00 20 00 41 00 cd 00  cc 00 cc 00 3e 00 00 00  |.. .A.......>...|
0007f180  00 00 a0 00 40 00 0a 00  d7 00 a3 00 3c 00 00 00  |....@.......<...|
0007f190  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
0007f1a0  00 00 00 00 00 00 01 00  00 00 00 00 00 00 00 00  |................|
0007f1b0  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
0007f1c0  03 00 03 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
0007f1d0  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
*
0007f200  00 00 00 00 00 00 00 00  00 00 00 00 8f 00 c2 00  |................|
0007f210  75 00 bf 00 00 00 00 00  00 00 00 00 00 00 00 00  |u...............|
0007f220  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
*
0007f250  00 00 00 00 00 00 00 00  00 00 00 00 04 00 04 00  |................|
0007f260  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
*
0007f280  00 00 00 00 c0 00 7f 00  00 00 00 00 c0 00 7f 00  |................|
*
0007f300  00 00 ff 00 0a 00 00 00  5a 00 00 00 00 00 c8 00  |........Z.......|
0007f310  00 00 fa 00 00 00 3c 00  00 00 64 00 00 00 00 00  |......<...d.....|
0007f320  00 00 8f 00 c2 00 89 00  41 00 0a 00 d7 00 23 00  |........A.....#.|
0007f330  3e 00 01 00 f0 00 e7 00  43 00 00 00 00 00 80 00  |>.......C.......|
0007f340  3f 00 14 00 00 00 e1 00  3a 00 a9 00 43 00 d3 00  |?.......:...C...|
0007f350  bc 00 23 00 41 00 00 00  ea 00 2e 00 45 00 20 00  |..#.A.......E. .|
0007f360  00 00 01 00 00 00 00 00  40 00 40 00 00 00 00 00  |........@.@.....|
0007f370  34 00 42 00 00 00 00 00  00 00 00 00 00 00 00 00  |4.B.............|
0007f380  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
0007f390  50 00 41 00 00 00 00 00  00 00 00 00 00 00 00 00  |P.A.............|
0007f3a0  00 00 41 00 00 00 00 00  00 00 00 00 e0 00 3f 00  |..A...........?.|
0007f3b0  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
*
0007f3e0  00 00 00 00 64 00 00 00  00 00 00 00 64 00 00 00  |....d.......d...|
0007f3f0  00 00 00 00 03 00 00 00  00 00 00 00 64 00 00 00  |............d...|
0007f400  00 00 00 00 64 00 00 00  00 00 00 00 03 00 00 00  |....d...........|
0007f410  00 00 00 00 03 00 00 00  00 00 00 00 1e 00 00 00  |................|
0007f420  00 00 00 00 1e 00 00 00  00 00 00 00 1e 00 00 00  |................|
*
0007f440  00 00 00 00 1e 00 00 00  00 00 00 00 00 00 00 00  |................|
0007f450  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
*
0007f580  00 00 00 00 00 00 c8 00  42 00 00 00 00 00 00 00  |........B.......|
0007f590  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
*
0007f5c0  40 00 40 00 ff ff ff ff  ff ff ff ff ff ff ff ff  |@.@.............|
0007f5d0  ff ff ff ff ff ff ff ff  ff ff ff ff ff ff ff ff  |................|
*
00080000

```

File: Marlin/src/module/configuration_store.cpp (hunk 4 above)
Comment (tpruvot) [reply Marlin/src/module/configuration_store.cpp]: The problem changing this file here is you are changing a shared code (with all platforms) just to fix a debug ability on an optional feature of the F1

**Comment by Msq001** [comment]
@tpruvot @tpruvot  Actually, https://github.com/MarlinFirmware/Marlin/blob/a0d782e46e8bffbdaea450c31707e9dbf0698fa8/Marlin/src/HAL/HAL_STM32F1/persistent_store_flash.cpp#L90 this file is unreasonable, at least in conflict with the configuration_store.cpp file.
for example:
EEPROM_WRITE(uint8_t); 
In fact, only one byte was written, but “pos += size + odd; ” pos += 1 + 1; pos increased 2.
This results in the final value of "eeprom_index" that is not equal to "sizeof (Settings Data)"
in this end， I get 
> m500
echo:Settings Stored (680 bytes; crc 16764)
Error:EEPROM datasize error.
[ERROR] Error:EEPROM datasize error.

> m504
SENDING:M504
Error:EEPROM datasize error.
[ERROR] Error:EEPROM datasize error.

echo:Index: 680 Size: 666

https://github.com/tpruvot/Marlin/commit/1bc5c97b95ac3f265b291955ecebb58886613f57#r34597762

**Comment by Msq001** [comment]
https://github.com/tpruvot/Marlin/commit/1bc5c97b95ac3f265b291955ecebb58886613f57#r34597762

**Comment by trouch** [comment]
@Msq001 yup
this PR get rid of all that

**Comment by thinkyhead** [comment]
> Ok, was able to make it work on the U20... and the SKRmini
> Here is the updated commit : tpruvot@1bc5c97

Note that you can submit a PR to the `trouch/Marlin` branch `fix/stm32-flash-eeprom-cleanup` (which this PR is based on) and when that PR is accepted, it will add your commits to this PR.

**Comment by thinkyhead** [comment]
> As a later improvement, we could add a flag to switch between pages at each eeprom save.

Anything that reduces wear is good!

**Comment by tpruvot** [comment]
i dont think its required, 10000 writes / 365 make a lot of years ^^

**Comment by tpruvot** [comment]
oh btw... #14877 for the tests

**Comment by tpruvot** [comment]
what about PRINTCOUNTER stats, its in config too ?

File: Marlin/src/module/configuration_store.cpp [hunk 6]
```diff
@@ -447,36 +447,19 @@ void MarlinSettings::postprocess() {
...
-  #else
-    #define EEPROM_SKIP(VAR) (eeprom_index += sizeof(VAR))
-  #endif
-
   #define EEPROM_START()          if (!persistentStore.access_start()) { SERIAL_ECHO_MSG("No EEPROM."); return false; } \
                                   int eeprom_index = EEPROM_OFFSET
   #define EEPROM_FINISH()         persistentStore.access_finish()
-  #define EEPROM_WRITE(VAR)       do{ persistentStore.write_data(eeprom_index, (uint8_t*)&VAR, sizeof(VAR), &working_crc);              UPDATE_TEST_INDEX(VAR); }while(0)
-  #define EEPROM_READ(VAR)        do{ persistentStore.read_data(eeprom_index, (uint8_t*)&VAR, sizeof(VAR), &working_crc, !validating);  UPDATE_TEST_INDEX(VAR); }while(0)
-  #define EEPROM_READ_ALWAYS(VAR) do{ persistentStore.read_data(eeprom_index, (uint8_t*)&VAR, sizeof(VAR), &working_crc);               UPDATE_TEST_INDEX(VAR); }while(0)
```
Comment (thinkyhead) [inline Marlin/src/module/configuration_store.cpp]: Looks to me like the original bug is caused by the `eeprom_index` variable not getting advanced in these three macros. Only `test_index` is being advanced.

**Comment by thinkyhead** [comment]
Another approach to this would be to simply change `PersistentStore::write_data` and `PersistentStore::read_data` so that they return the number of bytes actually written, and then these methods can take care of doing the extra padding. In addition, `PersistentStore::skip_data` should be added to advance the passed EEPROM address ref (`int &pos`). The `test_index` will then always exist with `DEBUG_EEPROM_READWRITE` enabled, always advanced by the size of the un-padded data.

**Comment by thinkyhead** [comment]
> oh btw... #14877 for the tests

Not needed, since the STM32 patch for SoftwareSerial is now merged.

**Comment by tpruvot** [comment]
> > oh btw... #14877 for the tests
> 
> Not needed, since the STM32 patch for SoftwareSerial is now merged.

euh, flash eeprom has nothing to do with tmc serial :p

**Comment by thinkyhead** [comment]
I thought you were making a separate branch because Travis has been choking on STM32. But now I see it was due to being the LONGER pins file, not CHEETAH.

**Comment by thinkyhead** [comment]
> what about PRINTCOUNTER stats, its in config too ?

Yes. Meanwhile, UBL meshes also have a separate EEPROM section.

**Comment by trouch** [comment]
> Another approach to this would be to simply change `PersistentStore::write_data` and `PersistentStore::read_data` so that they return the number of bytes actually written, and then these methods can take care of doing the extra padding. In addition, `PersistentStore::skip_data` should be added to advance the passed EEPROM address ref (`int &pos`). The `test_index` will then always exist with `DEBUG_EEPROM_READWRITE` enabled, always advanced by the size of the un-padded data.

What about the comparison between test_index and  `uint16_t MarlinSettings::datasize() { return sizeof(SettingsData); }` ?

**Comment by thinkyhead** [comment]
> What about the comparison between test_index 

If the `test_index` is being consistently advanced by the un-padded data size in all cases, then it should match the byte-packed `sizeof(SettingsData)` at the end. If `datasize` ought to return the padded data size, then we should change it. Or, we can add a `padded_datasize` method which would return the final padded `eeprom_index` value, then use `padded_datasize` where appropriate.

**Comment by thinkyhead** [comment]
I guess I'm thinking mainly about future-proofing, in that this PR is actually simplifying the flash EEPROM by removing the extra padding.

If we can ensure no padded eeproms will be coming along in the future, then I won't worry to much about it.

**Pull request merged**